doc/design.md

```diff
@@ -0,0 +1,316 @@
+the Cgo wrapper `RandN` returns a C string representing the possible C++
+exception.  The last parameter of `RanN` is the returned tensor.
+
+The Go function `gotorch.RandN` calls the wrapper `RandN` and converts the
+returned C string, if not null, into a Go panic by calling `MustNil`.
+
+```go
+func RandN(shape []int64, requiresGrad bool) Tensor {
+	rg := 0
+	if requiresGrad {
+		rg = 1
+	}
+	var t C.Tensor
+	MustNil(unsafe.Pointer(C.RandN((*C.int64_t)(unsafe.Pointer(&shape[0])),
+		C.int64_t(len(shape)), C.int64_t(rg), &t)))
+	SetTensorFinalizer((*unsafe.Pointer)(&t))
+	return Tensor{(*unsafe.Pointer)(&t)}
+}
+```
+
+## Porting Functionals
+
+The Python package torch.nn.functional provides functions defined as compounds
+of native functions. Generally, each Python function corresponds to a C++
+function in namespace `torch::nn::functional`. So we need to expose the C++
+function via Cgo, and then define a Go function that calls the Cgo function.
+
+Let’s take `torch.nn.functional.linear` as an example. The first step is to
+expose the C++ function via Cgo. To do that, we define a C/C++ function in
+`gotorch/cgotorch/functionals.cc` as the wrapper of the C++ function in
+libtorch.
+
+```c++
+const char *Linear(Tensor input, Tensor weight, Tensor bias, Tensor *result) {
+  try {
+    auto out = torch::linear(*input, *weight, (bias ? *bias : torch::Tensor()));
+    *result = new at::Tensor(out);
+    return nullptr;
+  } catch (const std::exception &e) {
+    return exception_str(e.what());
+  }
+}
+```
+
+The next step is to define a Go function in `gotorch/nn/functional/*.go` to call
+the above wrapper via Cgo.
+
+```go
+func Linear(input, weight, bias torch.Tensor) torch.Tensor {
+	var t C.Tensor
+	var cBias C.Tensor
+	if bias.T != nil {
+		cBias = C.Tensor(*bias.T)
+	}
+	torch.MustNil(unsafe.Pointer(C.Linear(
+		C.Tensor(*input.T),
+		C.Tensor(*weight.T), cBias, &t)))
+	torch.SetTensorFinalizer((*unsafe.Pointer)(&t))
+	return torch.Tensor{(*unsafe.Pointer)(&t)}
+}
+```
+
+Now, we can use the function ported from `torch.nn.functional` to
+`gotorch/nn/functional` in our application programs.
+
+```go
+import torch "github.com/wangkuiyi/gotorch"
+import F "github.com/wangkuiyi/gotorch/nn/functional"
+
+input := torch.RandN([]int64{32, 100}, false)
+weight := torch.RandN([]int64{100, 10}, true)
+out := F.Linear(input, weight, torch.Tensor{})
+```
+
+## Porting Modules
+
+Users can define PyTorch modules in Python and C++, but mostly in Python.
+
+To define a new module in Python, users derive a class from `torch.nn.Module`,
+adding data members corresponding to parameters, buffers, and sub-modules, and
+attaching the method `forward` that defines the forward pass of the module.
+
+The Python module-definition framework have Python decorators that call C++ core
+functions `Module::register_{parameter|buffer|module}`.  This registration
+mechanism is necessary as later users might call the module class’ methods like
```

Comment:
The Python module-definition framework and C++ module-definition framework are completely independent. Python module-definition framework does not call C++ `register_parameter
|buffer` functions.

In C++, users have to call `Module::register_parameter|buffer|module` explicitly. Following is an example of defining a Linear module.

```cpp
void LinearImpl::reset() {
  weight = register_parameter("weight",
    torch::empty({options.out_features(), options.in_features()}));
  if (options.bias()) {
    bias = register_parameter("bias", torch::empty(options.out_features()));
  } else {
    bias = register_parameter("bias", {}, /*requires_grad=*/false);
  }
  reset_parameters();
}
```

In Python, [`register_parameter`](https://github.com/pytorch/pytorch/blob/v1.6.0/torch/nn/modules/module.py#L289) method is hide behind [`__setattr__`](https://github.com/pytorch/pytorch/blob/v1.6.0/torch/nn/modules/module.py#L774) method. So, when adding a data member to the module class, the `register_parameter` in Python will be called implicitly.

```python
def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
        super(Linear, self).__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.weight = Parameter(torch.Tensor(out_features, in_features))
        if bias:
            self.bias = Parameter(torch.Tensor(out_features))
        else:
            self.register_parameter('bias', None)
        self.reset_parameters()
```

Reply:
Thank you for the very comprehensive examples. I am restructuring the section, embedding them into the new version.